fastwarc/fastwarc/cli.py

```diff
@@ -167,13 +169,87 @@ def check(infile, decompress_alg, verify_payloads, quiet, output):
                     click.echo(rec)
                 sys.exit(1)
 
+@main.command()
+@click.argument('infile', type=click.Path(dir_okay=False, exists=True))
+@click.argument('offset', type=click.INT)
+@click.option('--payload', is_flag=True,
+              help='output only record payload (transfer and/or content encoding are preserved')
+@click.option('--headers', is_flag=True, help='output only record (and HTTP) headers')
+def extract(infile, offset, payload, headers):
+    """Extract WARC record by offset.
+    """
+    compl_record = not (payload or headers)
+    with open(infile, 'rb') as stream:
+        stream.seek(offset)
+        record = next(iter(ArchiveIterator(stream)))
+        if headers or compl_record:
+            ostream = PythonIOStreamAdapter(sys.stdout.buffer)
+            record.headers.write(ostream)
+            sys.stdout.buffer.write(b'\r\n')
+            if record.is_http:
+                record.http_headers.write(ostream)
+                sys.stdout.buffer.write(b'\r\n')
+        if payload or compl_record:
+            buf = record.reader.read(4096)
+            while buf:
+                sys.stdout.buffer.write(buf)
+                buf = record.reader.read(4096)
+
 
 @main.group()
 def benchmark():
     """Benchmark FastWARC performance."""
     return 0
 
 
+def _index_record(output, fields, record, next_record_offset, file_name):
+    idx = OrderedDict()
+    for f in fields:
+        v = None
+        if f == 'offset':
+            v = str(record.stream_pos)
+        elif f == 'length':
+            v = str(next_record_offset - record.stream_pos)
+        elif f == 'filename':
+            v = file_name
+        elif f == 'http:status':
+            if (record.record_type & (WarcRecordType.response | WarcRecordType.revisit)) and record.http_headers:
+                idx[f] = str(record.http_headers.status_code)
+        elif f.startswith('http:'):
+            if record.http_headers:
+                v = record.http_headers.get(f[5:])
+        elif f == 'rec.content_length':
+            v = record.content_length
+        else:
+            v = record.headers.get(f)
+        if v is not None:
+            idx[f] = v
+    output.write(json.dumps(idx) + '\n')
+
+
+@main.command()
+@click.argument('infiles', type=click.Path(dir_okay=False, exists=True), nargs=-1)
+@click.option('-o', '--output', type=click.File('w'), default=sys.stdout,
+              help='Output file, default is stdout')
+@click.option('-f', '--fields', type=click.STRING,
+              default='offset,warc-type,warc-target-uri',
+              help='comma-separated list of indexed fields, eg. "offset", "length", "filename", '
+                   '"http:status", "http:<http-header>", or "<warc-record-header>"')
```

Comment:
Uppercase first letters (see above) and `click.STRING` should simply be `str`.
You may also want to set `show_default=True` here.